stp/stp.zkasm

```diff
@@ -0,0 +1,163 @@
+;; account for a call instruction is warm (or not).
+fn stp(INST=0xf0 u8, GAS_ACTUAL u64, GAS_MXP u64, GAS u256, VALUE u256, EXISTS u1, WARM u1) -> (OOGX=1 u1, GAS_UPFRONT=32000 u64, GAS_OOP u64, GAS_STIPEND u64)
+;; PRE: GAS_UPFRONT calculation does not overflow.
+{
+  var L_gas_diff u64
+  var stipend, gas_extra u64
+  var z u12
+  var c u1
+  ;;
+  if INST == 0xf0 goto create
+  if INST == 0xf5 goto create
+  ;; ===========================================================================
+  ;; CALL / CALLCODE / DELEGATECALL / STATICCALL
+  ;; ===========================================================================
+  gas_extra, stipend = call_gas_extra(INST, VALUE, WARM, EXISTS)
+  ;; determine upfront cost for call
+  c, GAS_UPFRONT = GAS_MXP + gas_extra
+  ;; sanity check precondition
+  if c == 1 goto failed_pre
+  ;; check for out-of-gas (oogx)
+  L_gas_diff, OOGX = gas_out_of_pocket(GAS_ACTUAL,GAS_UPFRONT)
+  ;; calculate gas_minc
+  GAS_OOP = min256_64(GAS,L_gas_diff)
+  ;; handle oogx case
+  if OOGX == 1 goto call_oogx
+  GAS_STIPEND = stipend
+  return
+call_oogx:
+  GAS_STIPEND=0
+  return
+  ;; ===========================================================================
+  ;; CREATE / CREATE2
+  ;; ===========================================================================
+create:
+  ;; determine upfront cost for create
+  c, GAS_UPFRONT = GAS_MXP + 32000
+  ;; sanity check precondition
+  if c == 1 goto failed_pre
+  ;; calculate out-of-pocket cost
+  GAS_OOP, OOGX = gas_out_of_pocket(GAS_ACTUAL,GAS_UPFRONT)
+  GAS_STIPEND = 0
+  return
+failed_pre:
+  ;; precondition failure
+  fail
+}
+
+;; Determines extra gas costs for a call instruction (such as for
+;; transfering value) and computes the gas stipend for the caller
+;; context.  The logic of this function is essentially determined by
+;; the yellow paper.
+fn call_gas_extra(inst=0xf1 u8, value u256, warm u1, exists u1) -> (gas_extra=2600 u64, stipend u64)
+;; PRE: inst in {0xf1,0xf2,0xf4,0xf5]
+{
+  var gas_transfer u16
+  var gas_new_account u16
+  var gas_access u16
+  ;;
+  if inst == 0xf1 goto call_only
+  if inst == 0xf2 goto call_other
+  if inst == 0xf4 goto call_no_transfer
+  if inst == 0xfa goto call_no_transfer
+  fail
+call_only:
+  if value == 0 goto call_no_transfer
+  if exists == 1 goto call_other
+  stipend = 2300 ;; G_callstipend
+  gas_transfer = 9000 ;; G_callvalue
+  gas_new_account = 25000 ;; G_newaccount
+  goto call
+call_other:
+  if value == 0 goto call_no_transfer
+  stipend = 2300 ;; G_callstipend
+  gas_transfer = 9000 ;; G_callvalue
+  gas_new_account = 0
+  goto call
+call_no_transfer:
+  stipend = 0
+  gas_transfer = 0
+  gas_new_account = 0
+call:
+  ;; check for warm access
+  if warm == 0 goto call_cold
+  gas_access = 100 ;; G_warmaccess
+  goto call_cont
+call_cold:
+  gas_access = 2600 ;; G_coldaccountaccess
+call_cont:
+  gas_extra = gas_access + gas_transfer + gas_new_account
+  return
```

Comment:
This looks correct but I find the numerous `goto` statements confusing. Would it not be possible to compute the various terms `stipend`, `gas_transfer`, ... each in their own little subsection ? Would it negatively impact the weight of the constraints ?

Reply:
Does zkASM have a notion of (numerical) constants ? E.g. for the various gas constants.